subsys/logging/log_backend_rtt.c

```diff
@@ -139,45 +138,59 @@ static int log_backend_rtt_write_drop(void)
 	return 0;
 }
 
-static int log_backend_rtt_write_block(void)
+static void on_write(void)
 {
-	unsigned int ret;
-	*line_pos = '\r';
+	host_present = true;
 
-	RTT_LOCK();
-	ret = SEGGER_RTT_WriteSkipNoLock(CONFIG_LOG_BACKEND_RTT_BUFFER,
-					 line_buf, line_pos - line_buf + 1);
-	RTT_UNLOCK();
-
-	if (ret) {
-		log_backend_rtt_flush();
-		return 0;
+	if (panic_mode) {
+		while (SEGGER_RTT_HasDataUp(CONFIG_LOG_BACKEND_RTT_BUFFER)) {
+			/* Pend until data is fetched by the host. */
+		}
 	}
-	return 1;
 }
 
-static int log_backend_rtt_write(void)
+static void on_failed_write(int retry_cnt)
 {
-	if (IS_ENABLED(CONFIG_LOG_BACKEND_RTT_MODE_BLOCK)) {
-		return log_backend_rtt_write_block();
-	} else if (IS_ENABLED(CONFIG_LOG_BACKEND_RTT_MODE_DROP)) {
-		return log_backend_rtt_write_drop();
+	if (retry_cnt == 0) {
+		host_present = false;
+	} else if (panic_mode) {
+		k_busy_wait(1000*RETRY_DELAY_MS);
+	} else {
+		k_sleep(RETRY_DELAY_MS);
 	}
 }
 
-static int log_backend_rtt_panic(u8_t *data, size_t length)
+static int data_out_block_mode(u8_t *data, size_t length, void *ctx)
 {
-	unsigned int written;
+	int ret;
+	int retry_cnt = RETRY_CNT;
 
-	/* do not respect mutex, take it over */
-	written = SEGGER_RTT_WriteNoLock(CONFIG_LOG_BACKEND_RTT_BUFFER, data,
-					 length);
-	log_backend_rtt_flush();
-	return written;
-}
+	do {
+		if (!panic_mode) {
+			RTT_LOCK();
+		}
+
+		ret = SEGGER_RTT_WriteSkipNoLock(CONFIG_LOG_BACKEND_RTT_BUFFER,
```

Comment:
What if `panic_mode` changes during call to SEEGER?

Reply:
if call is interrupted by panic it will not return here.

Reply:
As far as I remember, @andrewboie suggested, that we should disabled panic if we entered recoverable fault (thread is terminated, but system is operational). Shouldn't we take under consideration such mode of operation?

Reply:
if we are entering recoverable fault we shouldn't enter panic. If we enter today then it's a bug. Panic should be used only for nonrecoverable situations.

Reply:
OK